### PR TITLE
[Verification] Refactors Chunk Data Pack

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR addresses comments on an earlier PR: 

- Refactors the chunk data pack by replacing the `CollectionID` with the actual `Collection`. This was suggested by @AlexHentschel to reduce the surface for data-inconsistency attacks from the execution node and also reduce code complexity, i.e., one less check at the verification node side, and hence less slashing challenge.
- Introduces `StoredChunkDataPack` data type as an internal data type for the `badger`. The difference between the `StoredChunkDataPack` and `ChunkDataPack` is the former keeps a `CollectionID` while the latter keeps the actual `Collection`. This was suggested by @m4ksio so that we can keep the `ChunkDataPack`s and `Collection`s in two separate databases without adding an extra overhead of storing the `Collection`s at two places. 
- Refactors the notion of _system_ `ChunkDataPack`  as the one that has both a `nil` value for `Collection` and `CollectionID`. This was suggested by @AlexHentschel aim at avoiding the utilization of `flow.ZeroID` to represent the state of an undefined entity. The reason is `flow.ZeroID` corresponds to a valid hash value in the hash function range, and hence it also represents entities that hash to `flow.ZeroID`, e.g., collections. This results in an accidental Byzantine behavior as we inject a collision into our hash function. However, the Golang already us with a distinct value for undefined notion which is `nil`.

As a side effect, these changes also save some bandwidth overhead (although very small) by (slightly) lightening the `ChunkDataResponse` message.

### CODE REVIEW

File: engine/execution/convert.go [hunk 1]
```diff
@@ -5,14 +5,15 @@ import (
...
+	storagemodel "github.com/onflow/flow-go/storage/model"
 )
 
 func GenerateExecutionResultAndChunkDataPacks(
 	prevResultId flow.Identifier,
 	startState flow.StateCommitment,
 	result *ComputationResult) (
 	endState flow.StateCommitment,
-	chdps []*flow.ChunkDataPack,
+	chdps []*storagemodel.StoredChunkDataPack,
```
Comment (m4ksio) [inline engine/execution/convert.go:16]: Do we want to keep `StoredChunkDataPack` here? 
My initial though was that we can simply keep using `ChunkDataPack` everywhere and only have `StoredChunkDataPack` as internal badger storage object

Comment (yhassanzadeh13) [reply engine/execution/convert.go:16]: Right, I am going to keep the `StoredChunkDataPack` as an internal badger object. This is just at the draft state to examine the surface of the change and make sure that it is not breaking any logic.

File: engine/execution/computation/execution_verification_test.go [hunk 2]
```diff
@@ -197,10 +197,6 @@ func executeBlockAndVerify(t *testing.T, txs [][]*flow.TransactionBody) *executi
 
 	for i, chunk := range er.Chunks {
 		isSystemChunk := i == er.Chunks.Len()-1
-		var collection flow.Collection
-		if !isSystemChunk {
-			collection = executableBlock.CompleteCollections[chdps[i].CollectionID].Collection()
-		}
```
Comment (yhassanzadeh13) [inline engine/execution/computation/execution_verification_test.go:203]: Collection is now part of the chunk data pack, which is prepared by the `ExecutionState, no need for a separate fetch.

File: engine/execution/convert.go [hunk 3]
```diff
@@ -96,7 +95,7 @@ func GenerateExecutionResultForBlock(
 // GenerateChunk creates a chunk from the provided computation data.
 func GenerateChunk(colIndex int,
 	startState, endState flow.StateCommitment,
-	colID, blockID, eventsCollection flow.Identifier, txNumber uint16) *flow.Chunk {
```
Comment (yhassanzadeh13) [inline engine/execution/convert.go:99]: unused parameter

File: engine/execution/provider/engine_test.go [hunk 4]
```diff
@@ -188,14 +168,8 @@ func TestProviderEngine_onChunkDataRequest(t *testing.T) {
...
 			Return(nil)
 
-		execState.
-			On("GetBlockIDByChunkID", chunkID).
-			Return(blockID, nil)
-		execState.
-			On("ChunkDataPackByChunkID", mock.Anything, chunkID).
-			Return(chunkDataPack, nil)
-
-		execState.On("GetCollection", chunkDataPack.CollectionID).Return(&collection, nil)
```
Comment (yhassanzadeh13) [inline engine/execution/provider/engine_test.go:198]: Retrieving and composing a `ChunkDataPack` happens in the `ExecutionState`, and the created `ChunkDataPack` also contains a `Collection`. So no more collection retrieval is needed.

File: engine/execution/state/state.go [hunk 5]
```diff
@@ -306,7 +307,28 @@ func (s *state) StateCommitmentByBlockID(ctx context.Context, blockID flow.Ident
...
+		// non-system chunks have a non-nil collection
+		collection, err := s.GetCollection(*storedChunkDataPack.CollectionID)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve collection for chunk data pack: %w", err)
+		}
+
+		chunkDataPack.Collection = collection
+	}
+
+	return chunkDataPack, nil
```
Comment (yhassanzadeh13) [inline engine/execution/state/state.go:331]: `ChunkDataPack` has a `Collection` field (instead of the older version `CollectionID`). However, on the execution node a `ChunkDataPack` is stored as the `badger` internal type of  `StoredChunkDataPack`. The difference between the `ChunkDataPack` and `StoredChunkDataPack` is the former has a `Collection` filed, but the latter only keeps the `CollectionID` and not the actual collection. This is to avoid redundant storage of the same `Collection` on the execution node at two different places, i.e., a collection is stored at `Collections` badger storage, hence, just knowing its identifier, one can look it up. 

One can see the `StoredChunkDataPack` as a cold-version persistent data type that is converted to the `ChunkDataPack` by fetching its `Collection` upon request.

File: engine/verification/fetcher/engine.go [hunk 6]
```diff
@@ -228,12 +228,18 @@ func (e *Engine) processAssignedChunk(chunk *flow.Chunk, result *flow.ExecutionR
 // HandleChunkDataPack is called by the chunk requester module everytime a new requested chunk data pack arrives.
 // The chunks are supposed to be deduplicated by the requester.
 // So invocation of this method indicates arrival of a distinct requested chunk.
-func (e *Engine) HandleChunkDataPack(originID flow.Identifier, chunkDataPack *flow.ChunkDataPack, collection *flow.Collection) {
+func (e *Engine) HandleChunkDataPack(originID flow.Identifier, chunkDataPack *flow.ChunkDataPack) {
```
Comment (yhassanzadeh13) [inline engine/verification/fetcher/engine.go:231]: Received `ChunkDataPack` already contains the `Collection`.

File: engine/verification/fetcher/engine.go [hunk 7]
```diff
@@ -393,55 +396,45 @@ func (e *Engine) validateChunkDataPack(chunkIndex uint64,
...
 	// 3. collection id must match
-	err := e.validateCollectionID(collection, chunkDataPack, result, chunk)
+	err := e.validateCollectionID(chunkDataPack, result, chunk)
 	if err != nil {
-		return fmt.Errorf("could not validate collection: %x, from sender ID: %x, block ID: %x, resultID: %x, chunk ID: %x",
-			collection.ID(),
-			senderID,
-			blockID,
-			result.ID(),
-			chunk.ID())
```
Comment (yhassanzadeh13) [inline engine/verification/fetcher/engine.go:403]: Redundant information: all these are appended to the error at the higher levels of these nested calls.

File: engine/verification/fetcher/engine_test.go [hunk 8]
```diff
@@ -281,30 +281,18 @@ func TestProcessAssignChunkSealedAfterRequest(t *testing.T) {
...
-				cdp.CollectionID = unittest.IdentifierFixture()
-			},
-			mockStateFunc: func(identity flow.Identity, state *protocol.State, blockID flow.Identifier) {
-				// mocks a valid execution node as originID
-				mockStateAtBlockIDForIdentities(state, blockID, flow.IdentityList{&identity})
-			},
-			msg: "conflicting-collection-ID-with-chunk-data-pack",
-		},
-		{
-			alterChunkDataResponse: func(cdp *flow.ChunkDataPack, coll *flow.Collection) {
```
Comment (yhassanzadeh13) [inline engine/verification/fetcher/engine_test.go:299]: We no longer need this test case as one of the purposes of this refactoring is to get rid of unnecessary edge cases like this.

File: engine/verification/utils/unittest/fixture.go [hunk 9]
```diff
@@ -334,92 +331,11 @@ func ExecutionResultFixture(t *testing.T, chunkCount int, chain flow.Chain, refB
...
 		SpockSecrets:   spockSecrets,
 	}
 }
 
-// LightExecutionResultFixture returns a light mocked version of execution result with an
-// execution receipt referencing the block/collections. In the light version of execution result,
-// everything is wired properly, but with the minimum viable content provided. This version is basically used
-// for profiling.
-// TODO: remove this once new architecture of verification node is in place.
-func LightExecutionResultFixture(chunkCount int) *CompleteExecutionReceipt {
```
Comment (yhassanzadeh13) [inline engine/verification/utils/unittest/fixture.go:348]: We no longer use this fixture (applying the `TODO`)

File: model/verification/verifiableChunkData.go [hunk 10]
```diff
@@ -11,7 +11,6 @@ type VerifiableChunkData struct {
 	Chunk             *flow.Chunk           // the chunk to be verified
 	Header            *flow.Header          // BlockHeader that contains this chunk
 	Result            *flow.ExecutionResult // execution result of this block
-	Collection        *flow.Collection      // collection corresponding to the chunk
```
Comment (yhassanzadeh13) [inline model/verification/verifiableChunkData.go:14]: This is already included in the `ChunkDataPack` field 👇 .

File: engine/verification/verifier/engine_test.go [hunk 11]
```diff
@@ -116,20 +114,6 @@ func (suite *VerifierEngineTestSuite) TestNewEngine() *verifier.Engine {
 
 }
 
-func (suite *VerifierEngineTestSuite) TestInvalidSender() {
```
Comment (yhassanzadeh13) [inline engine/verification/verifier/engine_test.go:119]: Verifier engine is only receiving input through `ProcessLocal` from other engines of the same verification node, hence, this test is no longer needed.

File: storage/badger/chunkDataPacks.go [hunk 12]
```diff
@@ -49,7 +50,8 @@ func NewChunkDataPacks(collector module.CacheMetrics, db *badger.DB, byChunkIDCa
 }
 
 func (ch *ChunkDataPacks) Store(c *flow.ChunkDataPack) error {
-	err := operation.RetryOnConflictTx(ch.db, transaction.Update, ch.byChunkIDCache.PutTx(c.ChunkID, c))
+	sc := toStoredChunkDataPack(c)
```
Comment (yhassanzadeh13) [inline storage/badger/chunkDataPacks.go:53]: chunk data pack is converted internally to a `StoredChunkDataPack` for maintaining on storage.

File: engine/verification/utils/unittest/fixture.go [hunk 13]
```diff
@@ -294,24 +288,28 @@ func ExecutionResultFixture(t *testing.T, chunkCount int, chain flow.Chain, refB
...
+			} else {
+				// generates chunk data pack fixture for system chunk
+				eventsHash, err := flow.EventsListHash(computationResult.Events[i])
+				require.NoError(t, err)
 
-			chunk := execution.GenerateChunk(i, startState, endState, collectionID, executableBlock.ID(), eventsHash, uint16(txNumber))
-			chunkDataPack := execution.GenerateChunkDataPack(chunk, collectionID, computationResult.Proofs[i])
+				chunk = execution.GenerateChunk(i, startState, endState, executableBlock.ID(), eventsHash, uint16(1))
+				chunkDataPack = execution.GenerateChunkDataPack(chunk.ID(), chunk.StartState, nil, computationResult.Proofs[i])
+			}
```
Comment (yhassanzadeh13) [inline engine/verification/utils/unittest/fixture.go:312]: Decoupling chunk data pack for system chunks from non-system chunks for better extensibility and maintainability.

File: engine/execution/provider/engine.go [hunk 14]
```diff
@@ -158,55 +158,44 @@ func (e *Engine) onChunkDataRequest(
 	// extracts list of verifier nodes id
 	chunkID := req.ChunkID
 
-	log := e.log.With().
+	lg := e.log.With().
```
Comment (m4ksio) [inline engine/execution/provider/engine.go:161]: I quite like `log` here

File: storage/badger/chunkDataPacks.go [hunk 15]
```diff
@@ -67,25 +69,42 @@ func (ch *ChunkDataPacks) Remove(chunkID flow.Identifier) error {
...
 	batch.OnSucceed(func() {
-		ch.byChunkIDCache.Insert(c.ChunkID, c)
+		ch.byChunkIDCache.Insert(sc.ChunkID, sc)
 	})
-	return operation.BatchInsertChunkDataPack(c)(writeBatch)
+	return operation.BatchInsertChunkDataPack(sc)(writeBatch)
 }
 
-func (ch *ChunkDataPacks) ByChunkID(chunkID flow.Identifier) (*flow.ChunkDataPack, error) {
+func (ch *ChunkDataPacks) ByChunkID(chunkID flow.Identifier) (*storagemodel.StoredChunkDataPack, error) {
```
Comment (m4ksio) [inline storage/badger/chunkDataPacks.go:80]: That's inconsistent - we `Store(c *flow.ChunkDataPack)` but retrieve `*storagemodel.StoredChunkDataPack`.
I think this should both just work on `flow.ChunkDataPack` and use `StoredChunkDataPack` internally, never leaking it outside of the package

**Review by m4ksio — COMMENTED** [review]
Nice change, just little inconsistency to be addressed.

File: storage/badger/chunkDataPacks.go (hunk 15 above)
Comment (yhassanzadeh13) [reply storage/badger/chunkDataPacks.go:80]: Agree, but the problem is in order to retrieve as `*flow.ChunkDataPack` we need to replace the `CollectionID` in the `StoredChunkDataPack` with the actual `Collection`. For that, we need to incorporate the `Collections` database into the `ChunkDataPacks` database as a member field, i.e., the `badger.ChunkDataPacks` would look like this:
```
type ChunkDataPacks struct {
	db             *badger.DB
	collections    *Collections
	byChunkIDCache *Cache
}
```

Not sure if we want to go in this direction with the extra level of dependency relations it adds, i.e., we then will both have `badger.Collections` as well as `badger.ChunkDataPacks` (that internally has `badger.Collections`) in some places. 

Just to highlight, a low-hanging fruit is to make `badger.ChunkDataPacks` both store and retrieve as `StoredChunkDataPack`, i.e., `Store(c *storagemodel.StoredChunkDataPack)`

Comment (m4ksio) [reply storage/badger/chunkDataPacks.go:80]: I see the concern, but we already do this - see https://github.com/onflow/flow-go/blob/2366de82cbc5e6c68eadbadbe246784c16fc46aa/storage/badger/blocks.go#L25 - `Blocks` depends on other data providers to serve full blocks.
As long as its all stay within the same package, I think it's OK pattern

File: engine/execution/provider/engine.go (hunk 14 above)
Comment (yhassanzadeh13) [reply engine/execution/provider/engine.go]: fixed in https://github.com/onflow/flow-go/pull/968/commits/7e777048874f17c042ada6867194b6e60d3555cd. The original refactoring reason was to avoid confusion with the package name `log`.

Comment (zhangchiqing) [reply engine/execution/provider/engine.go]: I actually like this change. 

I ran into a few problem that I used `log.Info().Msg("some log...")`, but forgot to define the log with `log := e.log.With()...`

Then go-fmt would automatically import zerolog:
```
	"github.com/rs/zerolog/log"
```

This caused the log to lose all the context.

using `lg` prevents such mistake. Because if you forgot to define `lg := e.log.With()...`, then compiler would error

File: storage/badger/chunkDataPacks.go (hunk 15 above)
Comment (yhassanzadeh13) [reply storage/badger/chunkDataPacks.go]: Fixed

File: engine/execution/provider/engine.go (hunk 14 above)
Comment (yhassanzadeh13) [reply engine/execution/provider/engine.go]: @m4ksio so can I revert it back according to Leo's comment?

Comment (m4ksio) [reply engine/execution/provider/engine.go]: If you think thats proper way, go ahead

File: engine/verification/fetcher/engine.go [hunk 16]
```diff
@@ -228,12 +228,19 @@ func (e *Engine) processAssignedChunk(chunk *flow.Chunk, result *flow.ExecutionR
...
-func (e *Engine) HandleChunkDataPack(originID flow.Identifier, chunkDataPack *flow.ChunkDataPack, collection *flow.Collection) {
+func (e *Engine) HandleChunkDataPack(originID flow.Identifier, chunkDataPack *flow.ChunkDataPack) {
 	lg := e.log.With().
 		Hex("origin_id", logging.ID(originID)).
-		Hex("collection_id", logging.ID(collection.ID())).
 		Hex("chunk_id", logging.ID(chunkDataPack.ChunkID)).
 		Logger()
+
+	if chunkDataPack.Collection != nil {
+		// non-system chunk data packs have non-nil collection
```
Comment (zhangchiqing) [inline engine/verification/fetcher/engine.go:238]: better to use else to indicate it's a system chunk in the log

File: storage/badger/chunkDataPacks.go [hunk 17]
```diff
@@ -67,25 +71,63 @@ func (ch *ChunkDataPacks) Remove(chunkID flow.Identifier) error {
...
+	chdp := &flow.ChunkDataPack{
+		ChunkID:    schdp.ChunkID,
+		StartState: schdp.StartState,
+		Proof:      schdp.Proof,
+	}
+
+	if schdp.CollectionID != nil {
+		collection, err := ch.collections.ByID(*schdp.CollectionID)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrive collection for stored chunk data pack: %w", err)
```
Comment (zhangchiqing) [inline storage/badger/chunkDataPacks.go:100]: better to include the collection id

File: storage/badger/model/storedChunkDataPack.go [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+// the collection on a secondary storage.
+type StoredChunkDataPack struct {
+	ChunkID      flow.Identifier
+	StartState   flow.StateCommitment
+	Proof        flow.StorageProof
+	CollectionID *flow.Identifier
+}
+
+func (s StoredChunkDataPack) ID() flow.Identifier {
+	return s.ChunkID
```
Comment (zhangchiqing) [inline storage/badger/model/storedChunkDataPack.go:18]: why is chunkid?

File: engine/verification/fetcher/engine.go (hunk 16 above)
Comment (yhassanzadeh13) [reply engine/verification/fetcher/engine.go:238]: fixed in https://github.com/onflow/flow-go/pull/968/commits/042ff8feb53bda3b4b218992de77378c91121272

File: storage/badger/chunkDataPacks.go (hunk 17 above)
Comment (yhassanzadeh13) [reply storage/badger/chunkDataPacks.go]: Fixed in https://github.com/onflow/flow-go/pull/968/commits/0a22d848633ab2ec942b347202d3e7b0fe40c7d9.

File: engine/execution/provider/engine.go (hunk 14 above)
Comment (yhassanzadeh13) [reply engine/execution/provider/engine.go:161]: refactored back in https://github.com/onflow/flow-go/pull/968/commits/ad14e2fb59a0478a4e30e91b65fee79a518d96e0.

File: storage/badger/model/storedChunkDataPack.go (hunk 18 above)
Comment (yhassanzadeh13) [reply storage/badger/model/storedChunkDataPack.go]: This was the residue of another version of this implementation. Thanks for pointing out, I removed these methods in https://github.com/onflow/flow-go/pull/968/commits/f02ea8c8745cafcb526ccfded79c9fb2df3cab60. 

In an earlier draft of this PR, I was using `StoredChunkDataPack` out of the `badger` database as well. So, in some cases `StoredChunkDataPack` would be utilized as an alternative to the `ChunkDataPack`, and since `ChunkDataPack` is an entity, hence `StoredChunkDataPack` must be. And the `ID` and `CheckSum` methods of `StoredChunkDataPack` hence implemented the same as `ChunkDataPack` (otherwise we could not use them interchangeably: 

https://github.com/onflow/flow-go/blob/a1bee171b3a909c1aab808a559c244a6533ccd29/model/flow/chunk.go#L40-L51

However, at the final version of this PR, we decided to keep `StoredChunkDataPack` as an internal model for the `badger` database, and not to use it anywhere else, hence degrading it from an entity to a simple data structure.

**Review by zhangchiqing — APPROVED** [review]
Nice refactor!

**Review by m4ksio — APPROVED** [review]
